pamu2fcfg/pamu2fcfg.c

```diff
@@ -203,24 +221,73 @@ static int make_cred(const struct args *args, const char *path, fido_dev_t *dev,
        r == FIDO_ERR_PIN_BLOCKED)) {
     n = snprintf(prompt, sizeof(prompt), "Enter PIN for %s: ", path);
     if (n < 0 || (size_t) n >= sizeof(prompt)) {
-      fprintf(stderr, "error: snprintf prompt");
-      return -1;
+      fprintf(stderr, "error: snprintf prompt\n");
+      goto err;
     }
-    if (!readpassphrase(prompt, pin, sizeof(pin), RPP_ECHO_OFF)) {
-      fprintf(stderr, "error: failed to read pin");
-      explicit_bzero(pin, sizeof(pin));
-      return -1;
+    if ((pin = malloc(BUFSIZE)) == NULL) {
+      fprintf(stderr, "error: malloc\n");
+      goto err;
+    }
+    if (!readpassphrase(prompt, pin, BUFSIZE, RPP_ECHO_OFF)) {
+      fprintf(stderr, "error: failed to read pin\n");
+      goto err;
     }
     r = fido_dev_make_cred(dev, cred, pin);
   }
-  explicit_bzero(pin, sizeof(pin));
 
   if (r != FIDO_OK) {
     fprintf(stderr, "error: fido_dev_make_cred (%d) %s\n", r, fido_strerr(r));
-    return -1;
+    goto err;
   }
 
-  return 0;
+  if (args->authtok) {
+    if (!readpassphrase("Enter auth token: ", authtok, sizeof(authtok),
+                        RPP_ECHO_OFF)) {
+      fprintf(stderr, "error: failed to read auth token\n");
+      goto err;
+    }
+
+    if (args->pam_userpresence == 0 && args->no_user_presence) {
+      fprintf(stderr, "error: user presence is required for auth token\n");
+      goto err;
+    }
+
+    if (args->pam_userverification == 1 || args->user_verification == 1) {
+      uv = FIDO_OPT_TRUE;
+    } else if (args->pam_userverification == 0)
+      uv = FIDO_OPT_FALSE;
+    else {
+      uv = FIDO_OPT_OMIT;
+    }
+
+    if (args->pam_pinverification == 1 || args->pin_verification == 1) {
+      if (pin == NULL) {
+        fprintf(stderr,
+                "error: pin verification is required but no pin is found\n");
+        goto err;
+      }
```

Comment:
I don't think this state can be completely avoided. We don't necessarily need a pin to make the credential if we have UV, but we must use the same authentication requirement to generate the hmac secret as PAM will eventually use.

I changed the code to ask for the pin here instead of throwing an error.

Reply:
To reduce the complexity of this function, would it make sense to require some form of user verification for the credential creation step if the PAM configuration uses some form of UV? We can then determine the UV requirements for the entire function once towards the start of it.